buildprocess/generateCatalogIndex.ts

```diff
@@ -1,99 +1,245 @@
+  - Note:
+    - loads are somewhat randomised across catalog, so you don't hit one server with many requests
+    - one load may not equal one request. some groups/references do not make network requests
+
+ * @param basicAuth basic auth token to add to requests which include `baseUrl` (or `proxy/`)
+ */
+export default async function generateCatalogIndex(
+  configUrl: string,
+  baseUrl: string,
+  speedString: string | undefined,
+  basicAuth: string | undefined
+) {
+  let speed = speedString ? parseFloat(speedString) : 1;
+  if (speed < 1) speed = 1;
 
   if (!configUrl || !baseUrl) {
-    console.error(`\nUSAGE: node ./build/generateCatalogIndex.js <config-url> <base-url>\n`);
+    console.error(
+      `\nUSAGE: node ./build/generateCatalogIndex.js <config-url> <base-url> <speed>\n`
+    );
     process.exit(1);
   }
 
   // Make sure baseURL has trailing slash
-  baseUrl = baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`
+  baseUrl = baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`;
 
   // Overwrite browser APIs (eg XMLHttpRequest and fetch)
-  require('jsdom-global')(undefined, {
+  require("jsdom-global")(undefined, {
     url: baseUrl
-  })
-  global.XMLHttpRequest = window.XMLHttpRequest
-  XMLHttpRequest = window.XMLHttpRequest
+  });
+
+  // Add basic auth token for all XMLHttpRequest/fetch that use baseUrl hostname
+  if (basicAuth) {
+    const baseHostname = URI(baseUrl).hostname();
+
+    // Adapted from https://stackoverflow.com/a/43875390
+    const open = window.XMLHttpRequest.prototype.open;
+    window.XMLHttpRequest.prototype.open = function(
+      method: string,
+      url: string
+    ) {
+      open.apply(this, arguments as any);
+
+      if (URI(url).hostname() === baseHostname || url.startsWith("proxy/")) {
+        this.setRequestHeader("authorization", `Basic ${basicAuth}`);
+      }
+
+      // // print failed URLs
+      // this.onloadend = req => {
+      //   if (this.status < 200 || this.status >= 300) {
+      //     console.log("\n\n\n");
+      //     console.log(this.responseURL);
+      //     console.log(URI(this.responseURL).hostname() === baseHostname);
+      //     console.log("\n\n\n");
+      //   }
+      // };
+    };
+
+    // Alternative approach using Resource instead of XMLHttpRequest (can be swapped over if needed)
+
+    // const load = (Resource as any)._Implementations.loadWithXhr;
+    // (Resource as any)._Implementations.loadWithXhr = function(...args: any) {
+    //   if (URI(args[0]).hostname() === baseHostname) {
+    //     if (!args[4]) {
+    //       args[4] = {};
+    //     }
+    //     args[4]["authorization"] = `Basic ${basicAuth}`;
+    //   }
+
+    //   load.apply(this, args);
+    // };
+
+    // A fun method to add Auth headers to all requests with baseUrl
+    const newFetch = (
+      input: RequestInfo,
+      init?: RequestInit
+    ): Promise<Response> => {
+      const url = typeof input === "string" ? input : input.url;
```

Comment:
I know this is experimental at this stage, but would be nice if we can at some point extract code for patching requests to a separate file for readability.

Reply:
Good idea! I've moved it to `buildprocess/patchNetworkRequests.ts`